In [11]:
import pandas as pd
%matplotlib inline   
import seaborn as sns
import random as random


sns.set_theme()
sns.set_context("notebook")
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [12]:
dtype_dict = {
    'SE_Number': 'str',
    'FarmName_Pseudo': 'str',
    'CalvingDate': 'str',
    'CalvingSireBullID': 'str',
    'CalvingEase': 'str',
    'BreedName': 'str',
    'BirthDate': 'str',
    'Mother': 'str',
    'Father': 'str',
    'YearSeason': 'str',
    'MeanTemperature': 'float',
    'MeanRelativeHumidity': 'float',
    'MeanTHI_adj': 'float',
    'HW': 'Int64',
    'cum_HW': 'Int64',
    'MaxTemp15Threshold': 'Int64'
}

# Load the data using the dtype_dict
data = pd.read_csv('../Data/MergedData/TheCalvingData.csv', dtype=dtype_dict)

# Convert the date columns to datetime
date_columns = [
    'CalvingDate',
    'BirthDate'
]

for column in date_columns:
    data[column] = pd.to_datetime(data[column], errors='coerce')

data.head()

SE_Number FarmName_Pseudo CalvingDate CalvingSireBullID CalvingEase  \
0  SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman          11   
1  SE-30dc5787-1389        a624fb9a  2022-04-11   22-8352 Ventura          11   
2  SE-30dc5787-1389        a624fb9a  2023-05-15   22-8352 Ventura          11   
3  SE-30dc5787-1396        a624fb9a  2022-08-31     3-7725 Attack          11   
4  SE-a624fb9a-1039        a624fb9a  2022-01-07     3-7725 Attack          12   

  YearSeason BreedName  BirthDate            Mother Father  MeanTemperature  \
0     2022-2    02 SLB 2013-07-12               NaN    NaN        12.987500   
1     2022-2    01 SRB 2015-01-27               NaN    NaN         4.208333   
2     2023-2    01 SRB 2015-01-27               NaN    NaN        15.045833   
3     2022-3    02 SLB 2015-03-20               NaN    NaN        11.041667   
4     2022-1    02 SLB 2010-09-23  SE-a624fb9a-0966    NaN       -10.733333   

   MeanRelativeHumidity  MeanTHI_adj  HW  cum_HW  MaxTemp15Threshold  
0              0.755542    55.218420   0       0                   1  
1              0.485667    45.716935   0       0                   0  
2              0.524333    60.559813   0       0                   1  
3              0.794792    53.536477   0       0                   0  
4              0.880792    17.822442   0       0                   0

In [13]:
dtype_dict = {
    'Date': 'str',
    'FarmName_Pseudo': 'str',
    'SE_Number': 'str',
    'Age': 'Int64',
    'DailyYield': 'float',
    'PreviousDailyYield': 'float',
    'DailyYieldChange': 'float',
    'DaysInMilk': 'Int64',
    'YearSeason': 'str',
    'LactationNumber': 'Int64',
    'ExpectedYield': 'float',
    'NormalizedDailyYield': 'float',
    'NormalizedDailyYieldChange': 'float',
    'HeatStress': 'Int64',
    'Temp15Threshold': 'Int64',
    'HW': 'Int64',
    'cum_HW': 'Int64',
    'MeanTemperature': 'float',
    'MeanTHI_adj': 'float'
}

milk_data = pd.read_csv('../Data/MergedData/MilkApproachYieldData.csv', dtype=dtype_dict)
milk_data['Date'] = pd.to_datetime(milk_data['Date'], format='%Y-%m-%d')
milk_data.head(-5)

Date FarmName_Pseudo           SE_Number   Age  LactationNumber  \
0      2022-01-01        a624fb9a    SE-064c0cec-1189  3095                7   
1      2022-01-02        a624fb9a    SE-064c0cec-1189  3096                7   
2      2022-01-03        a624fb9a    SE-064c0cec-1189  3097                7   
3      2022-01-04        a624fb9a    SE-064c0cec-1189  3098                7   
4      2022-01-05        a624fb9a    SE-064c0cec-1189  3099                7   
...           ...             ...                 ...   ...              ...   
487083 2023-06-03        f454e660  SE-fcdf259d-0044-0  4150               10   
487084 2023-06-04        f454e660  SE-fcdf259d-0044-0  4151               10   
487085 2023-06-05        f454e660  SE-fcdf259d-0044-0  4152               10   
487086 2023-06-06        f454e660  SE-fcdf259d-0044-0  4153               10   
487087 2023-06-07        f454e660  SE-fcdf259d-0044-0  4154               10   

        DaysInMilk YearSeason  DailyYield  PreviousDailyYield  \
0              191     2022-1       30.77                0.00   
1              192     2022-1       48.22               30.77   
2              193     2022-1       30.53               48.22   
3              194     2022-1       42.26               30.53   
4              195     2022-1       38.49               42.26   
...            ...        ...         ...                 ...   
487083         347     2023-3       12.67               15.75   
487084         348     2023-3       22.31               12.67   
487085         349     2023-3       12.84               22.31   
487086         350     2023-3        9.47               12.84   
487087         351     2023-3        8.97                9.47   

        DailyYieldChange  ExpectedYield  NormalizedDailyYield  \
0                   0.00      29.739372              1.034655   
1                  17.45      29.692059              1.624003   
2                 -17.69      29.644756              1.029862   
3                  11.73      29.597463              1.427825   
4                  -3.77      29.550181              1.302530   
...                  ...            ...                   ...   
487083             -3.08      11.811455              1.072687   
487084              9.64      11.750492              1.898644   
487085             -9.47      11.689810              1.098393   
487086             -3.37      11.629408              0.814315   
487087             -0.50      11.569286              0.775329   

        NormalizedDailyYieldChange  HeatStress  Temp15Threshold  HW  cum_HW  \
0                         0.000000           0                0   0       0   
1                         0.587699           0                0   0       0   
2                        -0.596733           0                0   0       0   
3                         0.396318           0                0   0       0   
4                        -0.127580           0                0   0       0   
...                            ...         ...              ...  ..     ...   
487083                   -0.260764           0                1   0       0   
487084                    0.820391           0                1   0       0   
487085                   -0.810107           0                1   0       0   
487086                   -0.289783           0                1   0       0   
487087                   -0.043218           0                1   0       0   

        MeanTemperature  MeanTHI_adj  
0             -3.025000    28.012944  
1             -0.279167    32.898193  
2              2.033333    36.760487  
3              0.066667    31.939524  
4             -3.700000    26.498206  
...                 ...          ...  
487083        12.666667    53.132530  
487084        13.079167    56.726870  
487085        14.237500    58.482418  
487086        15.345833    60.546358  
487087        15.645833    61.559237  

[487088 rows x 19 columns]

In [14]:
# Merge the data based on Date and FarmName_Pseudo
merged_data = pd.merge(data, milk_data[['Date', 'FarmName_Pseudo', 'HeatStress']],
                       left_on=['CalvingDate', 'FarmName_Pseudo'],
                       right_on=['Date', 'FarmName_Pseudo'],
                       how='left')

# Drop the extra 'Date' column from the merge
merged_data = merged_data.drop(columns=['Date'])

merged_data.head(-5)

SE_Number FarmName_Pseudo CalvingDate CalvingSireBullID  \
0       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
1       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
2       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
3       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
4       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
...                  ...             ...         ...               ...   
339705  SE-afdd9a78-1209        afdd9a78  2022-09-03           09-6247   
339706  SE-afdd9a78-1210        afdd9a78  2022-09-30           03-9916   
339707  SE-afdd9a78-1211        afdd9a78  2022-08-22           09-8824   
339708  SE-afdd9a78-1215        afdd9a78  2022-10-24           09-8817   
339709  SE-afdd9a78-1217        afdd9a78  2022-12-13           09-8824   

       CalvingEase YearSeason BreedName  BirthDate            Mother   Father  \
0               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
1               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
2               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
3               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
4               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
...            ...        ...       ...        ...               ...      ...   
339705          11     2022-4    01 SRB 2020-03-18  SE-afdd9a78-1138  09-6225   
339706          11     2022-4    01 SRB 2020-03-21  SE-afdd9a78-1139  09-6225   
339707          11     2022-3    01 SRB 2020-03-22  SE-afdd9a78-1106  09-6223   
339708          11     2022-4    01 SRB 2020-06-15  SE-afdd9a78-1099  03-9862   
339709          11     2022-1    01 SRB 2020-06-15  SE-afdd9a78-1087  09-6225   

        MeanTemperature  MeanRelativeHumidity  MeanTHI_adj  HW  cum_HW  \
0             12.987500              0.755542    55.218420   0       0   
1             12.987500              0.755542    55.218420   0       0   
2             12.987500              0.755542    55.218420   0       0   
3             12.987500              0.755542    55.218420   0       0   
4             12.987500              0.755542    55.218420   0       0   
...                 ...                   ...          ...  ..     ...   
339705         9.708333              0.975125    51.746774   0       0   
339706         7.700000              0.843458    48.796863   0       0   
339707        14.570833              0.769917    60.174253   1       0   
339708         6.975000              0.961333    42.482672   0       0   
339709       -10.441667              0.816000    17.458285   0       0   

        MaxTemp15Threshold  HeatStress  
0                        1           0  
1                        1           0  
2                        1           0  
3                        1           0  
4                        1           0  
...                    ...         ...  
339705                   0        <NA>  
339706                   0        <NA>  
339707                   1        <NA>  
339708                   0        <NA>  
339709                   0        <NA>  

[339710 rows x 17 columns]

In [15]:
# Define the function to identify heat stress based on weather conditions
def identify_weather_based_heat_stress(row):
    if (row['HW'] == 1 or 
        row['cum_HW'] > 0 or 
        row['MeanTHI_adj'] > 61 or
        row['MeanTemperature'] > 15):
        return 1
    return 0

# Apply the function to fill NaN values in the HeatStress column
merged_data['HeatStress'] = merged_data.apply(
    lambda row: identify_weather_based_heat_stress(row) if pd.isna(row['HeatStress']) else row['HeatStress'], axis=1)

merged_data.head(-5)

SE_Number FarmName_Pseudo CalvingDate CalvingSireBullID  \
0       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
1       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
2       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
3       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
4       SE-064c0cec-1189        a624fb9a  2022-05-25    9-4733 Fillman   
...                  ...             ...         ...               ...   
339705  SE-afdd9a78-1209        afdd9a78  2022-09-03           09-6247   
339706  SE-afdd9a78-1210        afdd9a78  2022-09-30           03-9916   
339707  SE-afdd9a78-1211        afdd9a78  2022-08-22           09-8824   
339708  SE-afdd9a78-1215        afdd9a78  2022-10-24           09-8817   
339709  SE-afdd9a78-1217        afdd9a78  2022-12-13           09-8824   

       CalvingEase YearSeason BreedName  BirthDate            Mother   Father  \
0               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
1               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
2               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
3               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
4               11     2022-2    02 SLB 2013-07-12               NaN      NaN   
...            ...        ...       ...        ...               ...      ...   
339705          11     2022-4    01 SRB 2020-03-18  SE-afdd9a78-1138  09-6225   
339706          11     2022-4    01 SRB 2020-03-21  SE-afdd9a78-1139  09-6225   
339707          11     2022-3    01 SRB 2020-03-22  SE-afdd9a78-1106  09-6223   
339708          11     2022-4    01 SRB 2020-06-15  SE-afdd9a78-1099  03-9862   
339709          11     2022-1    01 SRB 2020-06-15  SE-afdd9a78-1087  09-6225   

        MeanTemperature  MeanRelativeHumidity  MeanTHI_adj  HW  cum_HW  \
0             12.987500              0.755542    55.218420   0       0   
1             12.987500              0.755542    55.218420   0       0   
2             12.987500              0.755542    55.218420   0       0   
3             12.987500              0.755542    55.218420   0       0   
4             12.987500              0.755542    55.218420   0       0   
...                 ...                   ...          ...  ..     ...   
339705         9.708333              0.975125    51.746774   0       0   
339706         7.700000              0.843458    48.796863   0       0   
339707        14.570833              0.769917    60.174253   1       0   
339708         6.975000              0.961333    42.482672   0       0   
339709       -10.441667              0.816000    17.458285   0       0   

        MaxTemp15Threshold  HeatStress  
0                        1           0  
1                        1           0  
2                        1           0  
3                        1           0  
4                        1           0  
...                    ...         ...  
339705                   0           0  
339706                   0           0  
339707                   1           1  
339708                   0           0  
339709                   0           0  

[339710 rows x 17 columns]

In [16]:
# Save the reordered DataFrame to a CSV file
merged_data.to_csv('../Data/MergedData/HeatStressCalvingData.csv', index=False)

### Variables Explanation for `HeatStressCalvingData.csv`

1. **SE_Number**:
   - Description: A unique identifier for each cow.
   - Datatype: `str`
   - Example: `SE-064c0cec-1189`

2. **FarmName_Pseudo**:
   - Description: A pseudonym for the farm where the data was collected.
   - Datatype: `str`
   - Example: `a624fb9a`

3. **CalvingDate**:
   - Description: The date when the cow calved.
   - Datatype: `str` (should be converted to `datetime` for operations)
   - Format: `YYYY-MM-DD`
   - Example: `2022-05-25`

4. **CalvingSireBullID**:
   - Description: The identifier for the bull used during the calving.
   - Datatype: `str`
   - Example: `9-4733 Fillman`

5. **CalvingEase**:
   - Description: A categorical variable indicating the ease of calving.
   - Datatype: `str`
   - Example: `11` (Easy, without assistance)

6. **BreedName**:
   - Description: The breed of the cow.
   - Datatype: `str`
   - Example: `02 SLB`

7. **BirthDate**:
   - Description: The birth date of the cow.
   - Datatype: `str` (should be converted to `datetime` for operations)
   - Format: `YYYY-MM-DD`
   - Example: `2013-07-12`

8. **Mother**:
   - Description: The identifier for the cow's mother.
   - Datatype: `str`
   - Example: `SE-a624fb9a-0966`

9. **Father**:
   - Description: The identifier for the cow's father.
   - Datatype: `str`
   - Example: `3-9071 Apollo`

10. **YearSeason**:
    - Description: The seasonal period based on the year and the month range.
    - Datatype: `str`
    - Example: `2022-2`
    - YearSeason parameters:
      - 1: Dec-Feb
      - 2: Mar-May
      - 3: Jun-Aug
      - 4: Sep-Nov

11. **MeanTemperature**:
    - Description: The mean temperature recorded on the calving day.
    - Datatype: `float`
    - Example: `12.987499999999999`

12. **MeanRelativeHumidity**:
    - Description: The mean relative humidity recorded on the calving day.
    - Datatype: `float`
    - Example: `0.7555416666666667`

13. **MeanTHI_adj**:
    - Description: The mean adjusted Temperature-Humidity Index for the calving day.
    - Datatype: `float`
    - Example: `55.21842`

14. **HW**:
    - Description: A binary variable indicating the presence of a heatwave on the calving day (0: no heatwave, 1: heatwave).
    - Datatype: `Int64`
    - Example: `0`

15. **cum_HW**:
    - Description: Cumulative number of heatwave days up to the calving day.
    - Datatype: `Int64`
    - Example: `0`

16. **MaxTemp15Threshold**:
    - Description: A binary variable indicating if the temperature exceeded 15°C on the calving day (0: no, 1: yes).
    - Datatype: `Int64`
    - Example: `1`

17. **HeatStress**:
    - Description: A binary variable indicating the presence of heat stress on the cow on the calving day, based on specific weather conditions.
    - Datatype: `Int64`
    - Example: `0`